pkg/util/kubernetes/autoscalers/datadogclient.go

```diff
@@ -0,0 +1,234 @@
+		appKey = config.SanitizeAPIKey(config.Datadog.GetString("app_key"))
+	}
+
+	// DATADOG_HOST used to be the only way to set the external metrics
+	// endpoint, so we need to keep backwards compatibility. In order of
+	// priority, we use:
+	//   - DD_EXTERNAL_METRICS_PROVIDER_ENDPOINT
+	//   - DATADOG_HOST
+	//   - DD_SITE
+	endpoint := os.Getenv("DATADOG_HOST")
+	if config.Datadog.GetString(metricsEndpointConfig) != "" || endpoint == "" {
+		endpoint = config.GetMainEndpoint(metricsEndpointPrefix, metricsEndpointConfig)
+	}
+
+	if appKey == "" || apiKey == "" {
+		return nil, errors.New("missing the api/app key pair to query Datadog")
+	}
+
+	log.Infof("Initialized the Datadog Client for HPA with endpoint %q", endpoint)
+
+	client := datadog.NewClient(apiKey, appKey)
+	client.HttpClient.Transport = httputils.CreateHTTPTransport()
+	client.RetryTimeout = 3 * time.Second
+	client.ExtraHeader["User-Agent"] = "Datadog-Cluster-Agent"
+	client.SetBaseUrl(endpoint)
+
+	return client, nil
+}
+
+type datadogIndividualClient struct {
+	client             *datadog.Client
+	lastQuerySucceeded bool
+	lastFailure        time.Time
+	lastSuccess        time.Time
+	retryInterval      time.Duration
+}
+
+const (
+	minRetryInterval = 30 * time.Second
+	maxRetryInterval = 30 * time.Minute
+)
+
+// DatadogFallbackClient represents a datadog client able to query metrics to a second Datadog endpoint if the first one fails
+type datadogFallbackClient struct {
+	clients        []*datadogIndividualClient
+	lastUsedClient int
+}
+
+// NewDatadogFallbackClient generates a new client able to query metrics to a second Datadog endpoint if the first one fails
+func newDatadogFallbackClient(endpoints []config.Endpoint) (*datadogFallbackClient, error) {
+	if len(endpoints) == 0 {
+		return nil, log.Errorf("external_metrics_provider.endpoints must be non-empty")
+	}
+
+	defaultClient, err := newDatadogSingleClient()
+	if err != nil {
+		return nil, err
+	}
+
+	ddFallbackClient := &datadogFallbackClient{
+		clients: []*datadogIndividualClient{
+			{
+				client:             defaultClient,
+				lastQuerySucceeded: true,
+				retryInterval:      minRetryInterval,
+			},
+		},
+	}
+	for _, endpoint := range endpoints {
+		client := datadog.NewClient(endpoint.APIKey, endpoint.APPKey)
+		client.HttpClient.Transport = httputils.CreateHTTPTransport()
+		client.RetryTimeout = 3 * time.Second
+		client.ExtraHeader["User-Agent"] = "Datadog-Cluster-Agent"
+		client.SetBaseUrl(endpoint.URL)
+		ddFallbackClient.clients = append(
+			ddFallbackClient.clients,
+			&datadogIndividualClient{
+				client:             client,
+				lastQuerySucceeded: true,
+				retryInterval:      minRetryInterval,
+			})
+	}
+
+	return ddFallbackClient, nil
+}
+
+func (ic *datadogIndividualClient) queryMetrics(from, to int64, query string) ([]datadog.Series, error) {
+	// TODO: check for err 500 (server side) instead of any kind of error because switching to DR doesn’t make sense if that’s the query which is bad.
```

Comment:
Should this TODO be addressed before merge?

Reply:
It’s unfortunately challenging because the used API returns an `error` which contains nothing more than a `string`: https://github.com/zorkian/go-datadog-api/blob/437d51d487bfc328fcedadef799fb92128bb2278/request.go#L84 So, we cannot know more precisely the type of the error without parsing the string.

Anyhow, we recently had an issue where a back-end problem led to “Invalid query” error.
In this incident, although the error was saying that the issue was coming from the query, switching to another datacenter would have helped (provided the BE bug wasn’t deployed everywhere).

In the light of this, it may finally make sense to switch to a different DC for any kind of reported error.